tracer/src/Datadog.Trace/AppSec/ApiSecurity.cs

```diff
@@ -6,44 +6,108 @@
 #nullable enable
 using System;
 using System.Collections.Generic;
-using System.IO;
-using System.IO.Compression;
-using System.Text;
-using Datadog.Trace.AppSec.Waf;
-using Datadog.Trace.Sampling;
-using Datadog.Trace.Vendors.Newtonsoft.Json;
-using Datadog.Trace.Vendors.Serilog;
+using System.Collections.Specialized;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
 
 namespace Datadog.Trace.AppSec;
 
 internal class ApiSecurity
 {
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<ApiSecurity>();
+    private readonly int _maxRoutesSize;
     private readonly bool _enabled;
-    private readonly OverheadController _overheadController;
+    private readonly TimeSpan _minTimeBetweenReprocessTimeSpan;
+    private readonly OrderedDictionary _processedRoutes = new();
+    private readonly Dictionary<string, bool> _apiSecurityAddress = new() { { "extract-schema", true } };
 
-    public ApiSecurity(SecuritySettings securitySettings)
+    public ApiSecurity(SecuritySettings securitySettings, int maxRouteSize = 4096)
     {
-        _overheadController = new(securitySettings.ApiSecurityMaxConcurrentRequests, (int)(securitySettings.ApiSecuritySampling * 100));
         // todo: later, will be enabled by default, depending on if Security is enabled
         _enabled = securitySettings.ApiSecurityEnabled;
+        _minTimeBetweenReprocessTimeSpan = TimeSpan.FromSeconds(securitySettings.ApiSecuritySampleDelay);
+        _maxRoutesSize = maxRouteSize;
     }
 
-    public bool TryTellWafToAnalyzeSchema(IDictionary<string, object> args)
+    public bool ShouldAnalyzeSchema(bool lastWafCall, Span localRootSpan, IDictionary<string, object> args, string? statusCode, IDictionary<string, object>? routeValues)
     {
-        if (_enabled && _overheadController.AcquireRequest())
+        try
         {
-            args.Add(AddressesConstants.WafContextProcessor, new Dictionary<string, bool> { { "extract-schema", true } });
-            return true;
-        }
+            var samplingPriority = localRootSpan.Context.TraceContext.SamplingPriority;
+            if (_enabled && lastWafCall && samplingPriority is SamplingPriorityValues.AutoKeep or SamplingPriorityValues.UserKeep)
```

Comment:
This is ok for now since we are still making the sampling decision when the first span starts. I am working on #5306 to implement the delayed trace sampling decision. We may need to add a line here later (depends on which PR is merged first 😅) to trigger a sampling decision if it hasn't happened yet. Something like:

```diff
-            var samplingPriority = localRootSpan.Context.TraceContext.SamplingPriority;
+            var samplingPriority = localRootSpan.Context.TraceContext.GetSamplingPriority(TriggerSamplingDecision.IfNotSet);

            if (_enabled && lastWafCall && samplingPriority is SamplingPriorityValues.AutoKeep or SamplingPriorityValues.UserKeep)
```